lib/Dialect/Arc/Transforms/MergeIfs.cpp

```diff
@@ -0,0 +1,367 @@
+    }
+
+    // Ensure we don't move past the move limit imposed by side effects.
+    earliest.maximize(moveLimit);
+    if (!earliest)
+      continue;
+
+    // Either move the op inside the single block that contains all uses, or
+    // move it to just before its earliest user.
+    if (allUsesInBlock && allUsesInBlock->getParentOp() == earliest.op) {
+      op.moveBefore(allUsesInBlock, allUsesInBlock->begin());
+      ++pass.numOpsSunk;
+      anyChanges = true;
+      LLVM_DEBUG(llvm::dbgs() << "- Sunk " << op << "\n");
+    } else {
+      // Insert above other ops that we have already moved to this earliest op.
+      // This ensures the original op order is maintained and we are not
+      // spuriously flipping ops around. This also works without the
+      // `insertionPoint` lookup, but can cause significant linear scanning to
+      // find the op before which we want to insert.
+      auto &insertionPoint = insertionPoints[earliest.op];
+      if (insertionPoint) {
+        auto order = opOrder.lookup(insertionPoint);
+        assert(order.first == earliest.order.first);
+        assert(order.second >= earliest.order.second);
+        earliest.op = insertionPoint;
+        earliest.order = order;
+      }
+      while (auto *prevOp = earliest.op->getPrevNode()) {
+        auto order = opOrder.lookup(prevOp);
+        if (order.first != earliest.order.first)
+          break;
+        assert(order.second > earliest.order.second);
+        earliest.op = prevOp;
+        earliest.order = order;
+      }
+      insertionPoint = earliest.op;
+
+      // Only move if the op isn't already in the right spot.
+      if (op.getNextNode() != earliest.op) {
+        LLVM_DEBUG(llvm::dbgs() << "- Moved " << op << "\n");
+        op.moveBefore(earliest.op);
+        ++pass.numOpsMovedToUser;
+        anyChanges = true;
+      }
+
+      // Update the current op's order to reflect where it has been inserted.
+      // This ensures that later moves to the same pile of moved ops do not
+      // reorder the operations.
+      order = earliest.order;
+      assert(order.second < unsigned(-1));
+      ++order.second;
+      opOrder[&op] = order;
+    }
+  }
+}
+
+void Sinker::mergeIfs() {
+  DenseSet<Value> writes, reads;
+
+  scf::IfOp lastOp;
+  for (auto ifOp : rootBlock.getOps<scf::IfOp>()) {
+    auto prevIfOp = std::exchange(lastOp, ifOp);
+    if (!prevIfOp)
+      continue;
+
+    // Only handle simple cases for now. (Same condition, no results, only
+    // single blocks, and both ifs either have or don't have an else region.)
+    if (ifOp.getCondition() != prevIfOp.getCondition())
+      continue;
+    if (ifOp.getNumResults() != 0 || prevIfOp.getNumResults() != 0)
+      continue;
+    if (!ifOp.getThenRegion().hasOneBlock() ||
+        !prevIfOp.getThenRegion().hasOneBlock())
+      continue;
+    if (!ifOp.getElseRegion().empty() &&
+        (!ifOp.getElseRegion().hasOneBlock() ||
+         !prevIfOp.getElseRegion().hasOneBlock()))
```

Comment:
What's the purpose of the `empty()` check here? If `hasOneBlock()` will error if the region is empty, can this not also happen here if `ifOp` doesn't have an empty else region but `prevIfOp` does?

Reply:
Oh nice catch! This was missing a check above that ensures that either both ops have an else region, or both ops don't have an else region. Once that's done, this check makes sure that _if_ they have a non-empty else region, the two regions must have exactly one block (a limitation of my implementation).

Reply:
You are absolutely right! That simplifies things.